Helm Release Support (#1023)
pulumi/pulumi-hugo: themes/default/content/docs/guides/adopting/from_kubernetes.md

<img src="/logos/tech/helm.svg" align="right" class="h-32 px-8 pb-4">

Pulumi supports two distinct means of using Helm Charts:

1. [Emulating Helm Charts to render resource templates](#emulating-helm-charts-with-chart-resources)
2. [Natively installing Helm Charts as Releases](#natively-installing-helm-charts-as-releases)

We discuss and provide examples for each approach in this section.

### Emulating Helm Charts With Chart Resources

> With the [Helm V2]({{< relref "/registry/packages/kubernetes/api-docs/helm/v2/release" >}}) and [Helm V3]({{< relref "/registry/packages/kubernetes/api-docs/helm/v3/release" >}}) chart resources, Pulumi renders the templates and applies them directly, much like with `ConfigFile` and `ConfigGroup` shown earlier, which means all provisioning happens client-side using your Kubernetes authentication setup without needing a server-side component such as Tiller (for Helm V2).

Reviewer: I wonder if we should drop mentions of Helm v2 here while we're at it

Author: Done.

> // Deploy the latest version of the stable/wordpress chart.

Reviewer: Several comments still refer to `stable/wordpress`

Author: Updated.

...
```

### Natively installing Helm Charts as Releases

> A new [Helm Release]({{< relref "/registry/packages/kubernetes/api-docs/helm/v3/release">}})) resource (GA as of [v3.15.0](https://github.com/pulumi/pulumi-kubernetes/releases/tag/v3.15.0) of the Pulumi Kubernetes Provider and SDK) is available in addition to the `Chart` resources discussed earlier. In this case, the Pulumi Kubernetes provider uses an embedded version of the Helm SDK to natively support managing [`Helm Releases`](https://helm.sh/docs/glossary/#release) on the target Kubernetes cluster. As a result, known limitations with the Chart resources around supporting [Helm Chart Lifecycle Hooks](https://helm.sh/docs/topics/charts_hooks/) are mitigated.

Reviewer: ```suggestion
The [Helm Release]({{< relref "/registry/packages/kubernetes/api-docs/helm/v3/release">}})) resource (GA as of [v3.15.0](https://github.com/pulumi/pulumi-kubernetes/releases/tag/v3.15.0) of the Pulumi Kubernetes Provider) is another option for installing Charts. In this case, the Pulumi Kubernetes provider uses an embedded version of the Helm SDK to provide full-fidelity support for managing [`Helm Releases`](https://helm.sh/docs/glossary/#release).
```

The `Release` resource type's inputs closely mirror the options supported by the Helm CLI and deviate slightly from the API supported by the `Chart` resources. Some key options are highlighted here:

> * `chart`: The required chart name (for instance, `"wordpress"`). In case of a local helm chart, a path can be specified as well.

Reviewer: ```suggestion
* `chart`: The required chart name (for instance, `"wordpress"`). For a local helm chart, a path can be specified instead.
```

* `repositoryOpts`: (Optional) Bag containing URL and authentication/authorization information for the hosting Helm repository, if any.
* `version`: (Optional) The semantic chart version to pull (by default `"latest"`).
* `values`: (Optional) A dictionary of named key/value values for Charts with parameters.
* `skipAwait`: (Optional) Whether or not to skip waiting on the availability of all resources installed by the chart. By default, this is set to `false` (i.e. awaits all resources) which allows us to chain dependent resources and execution to the `Release` resource.
* `timeout`: (Optional) When `skipAwait` is `false`, the amount of time to wait on resources being available. If the timeout expires, the release is marked as `failed`.

For more details on all the supported inputs, please refer to the [API reference documentation]({{< relref "/registry/packages/kubernetes/api-docs/helm/v3/release/#inputs" >}}).

> Unlike `Chart` resource types, `Release` doesn't include references to the underlying Kubernetes resources created during the installation. As a result, just the `Release` resource is encoded in Pulumi state by default. The `Release` resource type includes the [`ReleaseStatus`]({{< relref "/registry/packages/kubernetes/api-docs/helm/v3/release/#releasestatus" >}}) as an output type. The `namespace`, `name` and other fields in `ReleaseStatus` can be used to invoke `get` calls on relevant resource types to read installed resources into Pulumi state. See the example used in the next section for an instance of this.

Reviewer: ```suggestion
Unlike `Chart` resource types, `Release` doesn't include references to the underlying Kubernetes resources created during the installation. As a result, only the `Release` resource is stored in Pulumi state by default. The `Release` resource type includes the [`ReleaseStatus`]({{< relref "/registry/packages/kubernetes/api-docs/helm/v3/release/#releasestatus" >}}) as an output type. The following example shows how to retrieve resources managed by a Release.
```

#### Installing a Helm Release

> To illustrate provisioning a Helm Chart using Pulumi, we will deploy the same `wordpress` chart as we did using the Helm `Chart` using `Release` instead:

Reviewer: ```suggestion
We will deploy the same `wordpress` chart as before:
```

// Deploy the bitnami/wordpress chart.
const wordpress = new k8s.helm.v3.Release("wpdev", {
    chart: "wordpress",
    repositoryOpts: {
        repo: "https://charts.bitnami.com/bitnami",
    },
    version: "9.6.0",
});

// Get the status field from the wordpress service, and then grab a reference to the spec.
const svc = k8s.core.v1.Service.get("wpdev-wordpress", pulumi.interpolate`${wordpress.status.namespace}/${wordpress.status.name}-wordpress`);
// Export the ingress IP for the wordpress frontend.
export const frontendIp = svc.status.loadBalancer.ingress[0].ip;
```

{{% /choosable %}}

{{% choosable language python %}}

```python
import pulumi
from pulumi import Output
from pulumi_kubernetes.core.v1 import Service
from pulumi_kubernetes.helm.v3 import Release, ReleaseArgs, RepositoryOptsArgs

# Deploy the bitnami/wordpress chart.
wordpress = Release(
    "wpdev",
    ReleaseArgs(
        chart="wordpress",
        repository_opts=RepositoryOptsArgs(
            repo="https://charts.bitnami.com/bitnami",
        ),
        # Force to use ClusterIP so no assumptions on support for LBs etc. is required.
        version="9.6.0",
        values={
            "service": {

>                 "type": "ClusterIP",

Reviewer: This example doesn't match the others.

		frontendIp := pulumi.All(wordpress.Status.Namespace(), wordpress.Status.Name()).ApplyT(func(r interface{})(interface{}, error){
			arr := r.([]interface{})
			namespace := arr[0].(*string)
			name := arr[1].(*string)
			svc, err := corev1.GetService(ctx, "svc", pulumi.ID(fmt.Sprintf("%s/%s-wordpress", *namespace, *name)), nil)
			if err != nil {
				return "", nil
			}
			return svc.Status.LoadBalancer().Ingress().Index(pulumi.Int(0)).Ip(), nil

		})
		ctx.Export("frontendIp", frontendIp)

		if err != nil {
			return err
		}

		return nil
	})
}
```

{{% /choosable %}}

{{% choosable language csharp %}}

```csharp
using Pulumi;
using Pulumi.Kubernetes.Core.V1;
using System.Collections.Generic;
using Pulumi.Kubernetes.Types.Inputs.Helm.V3;
using Pulumi.Kubernetes.Helm.V3;

class MyStack : Stack
{
    public MyStack()
    {
        // Deploy the bitnami/wordpress chart.
        var wordpress = new Release("wpdev", new ReleaseArgs
        {
            Chart = "wordpress",
            RepositoryOpts = new RepositoryOptsArgs
            {
                Repo = "https://charts.bitnami.com/bitnami"
            },
            Version = "9.6.0",

>         });

Reviewer: This version doesn't include the ClusterIP config

Author: Removed from the python version instead.

        // Export the ingress IP for Wordpress frontend.
        this.FrontendIP = service.Status.Apply(status => status.LoadBalancer.Ingress[0].Ip);
    }

    [Output]
    public Output<string> FrontendIP { get; set; }
}
```

{{% /choosable %}}

{{< /chooser >}}
After running `pulumi up`, we will see the resulting resources created, and the load balanced IP address will be printed:

```
Updating (dev)

     Type                              Name                                      Status
 +   pulumi:pulumi:Stack               <project/stack>                           created
 +   ├─ kubernetes:helm.sh/v3:Release  wpdev                                     created
     └─ kubernetes:core/v1:Service     wpdev-wordpress

Outputs:
    frontendIp        : "34.71.25.45"

Resources:
    + 2 created

Duration: 1m28s
```

We can now use the Helm CLI to confirm that a release has been created:

```
helm list -a
NAME            NAMESPACE       REVISION        UPDATED                                 STATUS          CHART           APP VERSION
wpdev-xxxxxxxx  default         1               2022-02-01 01:06:08.513501 -0800 PST    deployed        wordpress-....
```

> In addition, unlike the Chart resources, existing Helm releases deployed using the Helm CLI can be imported readily using the `pulumi import` command.

Reviewer: ```suggestion
The Helm Release resource also supports importing existing Helm releases by using the `pulumi import` command.
```
